crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -5683,8 +5683,8 @@ mod tests {
             ",
         )?;
 
-        assert_public_ty(&db, "src/a.py", "e", "NameError");
-        assert_public_ty(&db, "src/a.py", "f", "error");
+        assert_public_ty(&db, "src/a.py", "e", "Unbound | NameError");
```

Comment:
That's correct, but it's actually a separate thing to anything that this PR is trying to fix. (Though we might not have an issue to track this detail, and we probably should!)

try/except blocks don't create new scopes in Python; most variables defined in exception handlers will persist until after the block has finished. Bindings created using `as` in the `except` itself are however special-cased by the interpreter — the interpreter literally inserts an implicit `del` statement for that specific variable prior to the block ending. This is to reduce the chance of creating reference cycles, which would otherwise be very common for this kind of pattern and would create a lot of unnecessary work for the garbage collector.

Reply:
I discussed this issue in more depth in the PR description to one of my previous PRs, and there's some discussion in that PR thread of how we plan to model this: https://github.com/astral-sh/ruff/pull/13267#issue-2510259051

Reply:
Created https://github.com/astral-sh/ruff/issues/13641 to track this.

Reply:
Should we add TODO comments for test assertions that we know should change in the future?

Reply:
> Should we add TODO comments for test assertions that we know should change in the future?

Now that we have `reveal_type` support, I think I can actually just fix them so that they don't need to change in this PR at all. I'll split that out into a separate PR.

Reply:
https://github.com/astral-sh/ruff/pull/13643